server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java

```diff
@@ -83,38 +83,42 @@ public LogicalPlan apply(JoinPlan plan,
         LogicalPlan lhs = resolvePlan.apply(plan.lhs());
         LogicalPlan rhs = resolvePlan.apply(plan.rhs());
 
-        if (lhs instanceof Collect == false || rhs instanceof Collect == false) {
-            return null;
-        }
-
         long lhsNumDocs = planStats.get(lhs).numDocs();
         long rhsNumDocs = planStats.get(rhs).numDocs();
 
         if (applyOptimization(lhsNumDocs, rhsNumDocs) == false) {
             return null;
         }
 
-        Collect smallerSide;
-        Collect largerSide;
+        LogicalPlan smallerSide;
+        LogicalPlan largerSide;
 
         boolean rhsIsLarger = rhsNumDocs > lhsNumDocs;
 
         if (rhsIsLarger) {
-            smallerSide = (Collect) lhs;
-            largerSide = (Collect) rhs;
+            smallerSide = lhs;
+            largerSide = rhs;
+        } else {
+            smallerSide = rhs;
+            largerSide = lhs;
+        }
+
+        Collect largerSideCollect;
+        if (largerSide instanceof Collect collect) {
+            largerSideCollect = collect;
+            smallerSide = planStats.memo().extract(smallerSide);
```

Comment:
The sub-query needs to be extracted and cannot contain group references because the memo cannot resolve it by the end of optimization cycle. The group-reference handling works over replace-sources.

Reply:
Isn't `resolvePlan.apply` taking care of this? I don't quite follow

Reply:
The subquery needs to be completely recursively resolved, so that there is not group reference left. `resolvePlan.apply` does not do that. It only resolved the current group reference to the referenced logical plan.

Reply:
It could done with `resolvePlan.apply` but this requires that it is called recursively on the sub-query until all sub-queries are resolved.